Add support for bech32cosmos

zmanian opened the pull request:
Implements #65 

Adds bech32cosmos as dependency in glide
Turn Address into a struct with an anonymous field to allow for method overrides
Added change String function to do bech32

zmanian committed:
Use Bech32Cosmos when converting  Address to/from JSON and to String
Use Bech32Cosmos when converting  PubKey to Strings

zmanian committed:
Ensure the go implementation of bech32cosmos is installed

Update the glide.lock

zmanian committed:
The bech32cosmos spec requires all values to be go-wire encoded before passed to the bech encoder

zmanian committed:
Test the roundtrip of to and from string address deserialization and fix resulting bugs

zmanian committed:
Add deserialization and round trips for pub keys. Remove padding that was breaking things

adrianbrink commented on pub_key.go:
```diff
@@ -4,6 +4,9 @@ import (
 	"fmt"
+	"strings"
+
+	"github.com/cosmos/bech32cosmos/go"
```
Collapse this into the other external imports

adrianbrink commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
+	if err != nil {
+		return err
+	}
+	if strings.ToUpper(readable) != "CSMSADDR" {
```
Mh, I'm wondering whether we want to enforce the address in go-crypto. I can imagine that other zones would like to use a very similar bech encoding scheme. We should allow users to define their own `readable`.

zmanian replied:
So where does the readable value get injected from?

ebuchman replied:
Think it makes sense to unmarshal the readable and the data into the `Address` struct. Then consumers of go-crypto can enforce what they need to about the addresses, and we can do things like NewAccountAddress and NewValidatorAddress

adrianbrink commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
+}
+
+func (a *Address) MarshalJSON() ([]byte, error) {
+
```
No line break

adrianbrink commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
 // Use an alias so Unmarshal methods (with ptr receivers) are available too.
-type Address = cmn.HexBytes
+type Address struct {
+	cmn.HexBytes
```
We should track the human readable part here.

adrianbrink commented on pub_key.go:
```diff
@@ -134,9 +240,41 @@ func (pubKey PubKeySecp256k1) VerifyBytes(msg []byte, sig_ Signature) bool {
+	}
+	conv, err := bech32cosmos.ConvertBits(marshaled, 8, 5, true)
+
+	bech, err := bech32cosmos.Encode("CSMSPUB", conv)
```
Same here. I don't think we should enforce the human-readable part.

ebuchman commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
+func (a *Address) String() string {
+	marshaled, err := cdc.MarshalBinary(a)
+	if err != nil {
+		return "go-wire err:" + err.Error()
```
panic ?

ebuchman commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
+	return nil
+}
+
+func (a *Address) MarshalJSON() ([]byte, error) {
```
can't this just call String()?

zmanian replied:
So I wanted to preserve the same API as `HexBytes` offered and I need an override here because `HexBytes` implements `MarshalJSON`/`UnmarshalJSON`

ebuchman commented on pub_key.go:
```diff
@@ -15,7 +18,77 @@ import (
+		return err
+	}
+
+	if err != nil {
```
extra err check ?

ebuchman commented on pub_key.go:
```diff
@@ -76,7 +149,40 @@ func (pubKey PubKeyEd25519) ToCurve25519() *[32]byte {
+	}
+	conv, err := bech32cosmos.ConvertBits(marshaled, 8, 5, true)
+
+	bech, err := bech32cosmos.Encode("CSMSPUB", conv)
```
lets use constants instead of the magic strings for things we don't give external control over.

Is that right? Addresses will be `CSMSACC` or  `CSMSVAL` but pubkeys will always be `CSMSPUB` ?

ebuchman commented on pub_key.go:
```diff
@@ -76,7 +149,40 @@ func (pubKey PubKeyEd25519) ToCurve25519() *[32]byte {
+		return err
+	}
+
+	if err != nil {
```
double err

ebuchman commented on pub_key_test.go:
```diff
@@ -10,16 +10,20 @@ import (
+		bech32pub:  "csmspub:ucfk4sszj58peh7tzv7kqfqsnl2gnae5a669qfqcu5uv9pyp7g4uufm0yjxqpj95k8",
 	},
 }
 
```
we should add some tests for invalid addrs/pubkeys

ebuchman commented on wire.go:
```diff
@@ -16,6 +16,8 @@ func init() {
 }
 
 func RegisterWire(cdc *wire.Codec) {
+	cdc.RegisterConcrete(Address{},
```
should we be prefix encoding addresses?

zmanian replied:
I don't know?

jaekwon replied:
No we shouldn't.  An address is just a []byte, happens to be hex encoded in JSON.  We don't want to register an Address, since it doesn't implement any interface (it's not 1 of anything).

ebuchman reviewed (commented):
Thanks Zaki!

zmanian committed:
Move readable into the address struct

zmanian committed:
Move the readable part of pubkeys into the struct

zmanian committed:
Correctly handle unmarshal error

zmanian committed:
Switch the pubkeys to use readable.

Potential API breakage cause PubKeys are now structs and this affect the go-wire encoding

Remove the base58 encoding from tests

zmanian committed:
Panic on errors when converting to string presentations

zmanian committed:
Add ed25519 tests

zmanian commented:
I'm having trouble thinking of good failure tests. Open to suggestions

zmanian committed:
Track readable for addresses in the struct

zmanian committed:
Verify that human readable part via the interface to keys and addresses

zmanian committed:
Update glide.lock

zmanian committed:
updated glibe.lock

zmanian commented:
As far as I can tell, this works with go-wire develop.

jaekwon commented:
Commenting on https://github.com/cosmos/cosmos-sdk/issues/263

ebuchman commented:
We should revive this and get it merged. Looks like it needs a rebase.

Other points:
- the same error is checked twice a couple times (see comments)
- dont amino register addresses

Also, do we need to change the types to structs just so we can memoize the humanreadable ? it might be nicer to just compute the human readable form whenever we need it and leave the type definitions as is. Though we'll probably need `type Address cmn.HexBytes` instead of `type Address = cmn.HexBytes` so we can override the JSON methods

b00f commented:
If I am not mistaken the bech32cosmos is an error correction and detection algorithm. Regarding to this ,  is it possible to check a given address and make sure it is derived from a public key?

zmanian commented:
@b00f Because an address is a digest("hash") of a public key, you can't check than an address was generated correctly unless you know the public key value already.

b00f commented:
As I discussed earlier with @ebuchman it's possible to check address without having public-key by providing checksum at the end of address.
Please check here:
https://github.com/tendermint/abci/pull/245#discussion_r190301914

>In go-crypto (which is a lovely project) it will be helpful if we can add checksum to the address object. This way addresses can be validated by themselves (like Bitcoin addresses), which would avoid much user error by avoiding sends to blackhole addresses ;) :
https://bitcoin.stackexchange.com/questions/32353/how-do-i-check-the-checksum-of-a-bitcoin-address

b00f commented:
It will be very useful, especially avoiding sending tokens to a mistakenly wrong typed address. Ethereum has been suffering from not having this checksum in its address.

zmanian commented:
@b00f Yes bech32cosmos adds checksums and will be adopted in the cosmos sdk

zmanian commented:
See https://github.com/cosmos/cosmos-sdk/pull/1029

b00f commented:
That will be great. May I know how we can check an address here in go-cypto?

zmanian commented:
So we decided not to have address verification in go-crypto and instead put it in the sdk

ebuchman commented:
> So we decided not to have address verification in go-crypto and instead put it in the sdk

I'm wondering if this was the right decision. People not using the SDK will probably want this. I guess the challenge was that its weird to have "COSMOSACCADDR" in the `tendermint/go-crypto` repo ?
